### DISCUSSION TITLE
New package for Ubuntu

### DISCUSSION THREAD

**Discussion by jhonny-oliveira** [fougue/mayo#221] [Q&A]
Hi!

I have packaged the latest version of Mayo for Ubuntu Jammy+ and would like to leverage it as part of [XtraDeb](https://xtradeb.net) initiative/PPAs.

Could you please review the control and the copyright files I have drafted?

<details>
  <summary>debian/control</summary>

````
Source: mayo
Section: graphics
Priority: optional
Maintainer: Jhonny Oliveira <jhonny.oliveira@gmail.com>
Build-Depends: debhelper-compat (= 13),
               dh-exec,
               help2man,
               libocct-data-exchange-dev,
               libocct-draw-dev,
               qtbase5-dev,
               libqt5svg5-dev
Standards-Version: 4.5.0
Homepage: https://github.com/fougue/mayo

Package: mayo
Architecture: any
Depends: ${misc:Depends}, ${shlibs:Depends}
Description: Opensource 3D CAD viewer and converter
 Mayo is an open-source 3D CAD viewer and converter based on Open CASCADE
 Technology and Qt.
 .
 Users can view and convert 3D CAD files in different formats (STEP, IGES, mesh
 formats, …).
 Mayo helps also user to graphically explore/analyze assembly trees and view
 properties (e.g. area, volume, ...).
````
</details>

<details>
  <summary>debian/copyright</summary>

```
Format: http://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
Upstream-Name: mayo
Source: https://github.com/fougue/mayo

Files: *
Copyright: Copyright (c) 2016, Fougue Ltd. <http://www.fougue.pro>
License: BSD-2-Clause

Files: debian/*
Copyright: Copyright (C) 2023 Jhonny Oliveira <jhonny.oliveira@gmail.com>
License: GPL-3.0+

License: BSD-2-Clause
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:
 ․
  1. Redistributions of source code must retain the above copyright notice,
     this list of conditions and the following disclaimer.
 ․
  2. Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in the
     documentation and/or other materials provided with the distribution.
 ․
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.

License: GPL-3.0+
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 .
 This package is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 .
 You should have received a copy of the GNU General Public License
 along with this program. If not, see <http://www.gnu.org/licenses/>.
 .
 On Debian systems, the complete text of the GNU General
 Public License version 3 can be found in "/usr/share/common-licenses/GPL-3".
```
</details>

Thank you in advance!
Jhonny Oliveira @ [XtraDeb](https://xtradeb.net)

**Comment by HuguesDelorme** [comment]
Hello and thanks for this initiative!
I'm wondering about the Mayo exe binary provided by the package. It is built with a specific version of OpenCascade
Let's assume Mayo exe was built against OpenCascade v7.7(latest). When the user will install Mayo package the OpenCascade runtime is not garanteed to be of the same version, eg Ubuntu 22.04 provides OpenCascade v7.4
The problem is that OpenCascade minor versions are not source and binary backward/forward compatible.
I have little knownledge about application Linux packaging, maybe this kind of problem is solved by the package system(eg a package can state a dependency to match an exact version otherwise it refuses to install)
Can you point me to the build or script files of the package you are preparing for Mayo?

About debian/control:
* notice for Linux system, due to OpenCascade limitation, Mayo works only with xcb(wayland isn't supported). Add xcb dependencies to `Build-Depends` section?
* `Architecture` should be probably more restrictive? Not sure if arm is supported

About debian/copyright:
* please change the date to `Copyright (c) 2023, Fougue Ltd. <https://www.fougue.pro>`
* please keep only the BSD-2-clause license(remove the GPL-3+ licensing)

**Reply by jhonny-oliveira** [reply]
Thank you for answering quickly!

Concerning your first question, I don't think there is a problem,  but I can't answer with 100% certainty. For instance, in this package's specific case, I had to build it for both Ubuntu versions. The package built in Jammy doesn't work in lunar due to missing dependency (library versions). So, it is always built against the matching distribution versions.

xcb is now added to the Build-Depends.

About the Architecture, I always do it this way for new packages. No complains this far, but I must confess, I only test them on amd64. Let me know if you want me to restrict it.

Your copyright is updated, but I left the GPL3+. They only concern the packaging files I created (debian/*). I usually publish my files under this license.

I did not publish anything anywhere yet. Wanted to check with you first.

For completion,

<details>
  <summary>debian/rules</summary>

  ```
#!/usr/bin/make -f

CASCADE_INC_DIR=`dpkg -L libocct-foundation-dev | grep -i "Standard_Version.hxx" | sed "s/\/Standard_Version.hxx//i"`

%:
    dh $@

override_dh_auto_configure:
    dh_auto_configure -- CASCADE_INC_DIR=$(CASCADE_INC_DIR)

execute_after_dh_auto_build:
    help2man -n "Opensource 3D CAD viewer and converter" -o mayo.1 $(CURDIR)/mayo
  ```
</details>

If you are OK with it, I will publish it in launchpad.net and you can review the complete package.

**Reply by HuguesDelorme** [reply]
That sounds fine to me, nevertheless I find the Linux "package" way less interesting than AppImage.
Mayo AppImage allows to use latest OpenCascade(7.7 as of today) even on Ubuntu versions that doesn't provide it(example Ubuntu Jammy provides OpenCascade 7.5)
So Mayo installed on Ubuntu Jammy with `apt-get install` won't be able to export to OBJ format or import VRML files.
It would be nice if you handle user issues about that package after it's made publicly available(any issue reported in Mayo GitHub issues related to the debian package will be forwarded to the launchpad space of your package)

**Reply by jhonny-oliveira** [reply]
You have a good point on OpenCascade. Maybe, I consider back porting it to Jammy and rebuild Mayo. No promises, though!

No worries! I intend to handle the issues concerning my packaging recipe ([bugs@xtradeb.net](https://bugs.launchpad.net/xtradeb)). Likewise, I will forward the remaining ;-).

The package is published: [Mayo@xtradeb.net](https://xtradeb.net/apps/mayo).

Feel free to review the entire package: [Mayo@launchpad.net](https://launchpad.net/~xtradeb/+archive/ubuntu/apps/+packages?field.name_filter=mayo). Your feedback is welcome!

Thank you!

**Comment by alexmyczko** [comment]
Meanwhile mayo landed in Debian and should get synced to Ubuntu soonish...

**Reply by HuguesDelorme** [reply]
@alexmyczko
Mayo build is failing on mips64el, riscv64, alpha, hppa, etc. because of fast_float 3rdparty library which doesn't detect the OS:
````
src/3rdparty/fast_float/float_common.h:20:2: error: #error Unknown platform (not 32-bit, not 64-bit?)
   20 | #error Unknown platform (not 32-bit, not 64-bit?)
````
I will raise an issue in fast_float repo for that, or if you want to do it then let me know.

Mayo build fails on armel because of weird linking error:
````
/usr/bin/ld: libMayoCore.a(task_manager.cpp.o): undefined reference to symbol '__atomic_fetch_add_8@@LIBATOMIC_1.0'
/usr/bin/ld: /lib/arm-linux-gnueabi/libatomic.so.1: error adding symbols: DSO missing from command line
````

Other point: do you know where I can find a download statistic for the Mayo debian package?

**Reply by alexmyczko** [reply]
no download statistic but installed stats on popcon.debian.org

please go ahead with the fixing

**Comment by alexmyczko** [comment]
if you have a linux screenshot, feel free to add it to: https://screenshots.debian.net/package/mayo
